universal-application-tool-0.0.1/app/assets/javascripts/validation.ts

```diff
@@ -0,0 +1,274 @@
+/** The validation controller provides basic client-side validation of form fields. */
+class ValidationController {
+  static readonly ADDRESS_QUESTION_CLASS = '.cf-question-address';
+  static readonly ENUMERATOR_QUESTION_CLASS = '.cf-question-enumerator';
+  static readonly NAME_QUESTION_CLASS = '.cf-question-name';
+
+  static readonly ENUMERATOR_DELETE_TEMPLATE = 'enumerator-delete-template';
+  static readonly BLOCK_SUBMIT_BUTTON_ID = 'cf-block-submit';
+
+  isAddressValid = true;
+  isEnumeratorValid = true;
+  isNameValid = true;
+
+  /** 
+   * Validating on input will call the validators whenever a field is updated.
+   * The default behavior is to only attempt to validate on submit.
+   * 
+   * Validate on input will also disable the submit button when errors are detected.
+   */
+  validateOnInput = false;
+
+  constructor() {
+    // attach listener to block form.
+    const blockForm = document.getElementById("cf-block-form");
+    if (blockForm) {
+      blockForm.addEventListener("submit", (event) => { return this.attemptSubmit(event); });
+    }
+
+    // Maybe attach listeners to all form fields.
+    this.addQuestionListeners();
+  }
+
+  /** attach listeners so that we know when to update validations. */
+  private addQuestionListeners() {
+    // Always at least add basic listener for enumerators.
+    this.addEnumeratorListeners();
+
+    if (this.validateOnInput) {
+      this.addAddressListeners();
+      this.addNameListeners();
+    }
+  }
+
+  private attemptSubmit(event: Event): boolean {
+    this.checkAllQuestionTypes();
+    if (!this.isValid()) {
+      event.preventDefault();

```

Comment:
(blocking) I just looked up what `return false` and `preventDefault` does. It seems like we'd want `event.preventDefault()` at the top of this function before `checkAllQuestionTypes` to prevent the default behavior from happening. I don't know if putting `event.preventDefault()` right next to `return false` does anything.

Can you clarify?

Reply:
`preventDefault` and `return false` are kind of redundant. It depends on the context of how the function was attached/called. So, I typically just do both for coverage.